couchdb/mango/query.go

```diff
@@ -0,0 +1,190 @@
+package mango
+
+import (
+	"encoding/json"
+	"unicode"
+)
+
+// This package provides utility structures to build mango queries
+
+////////////////////////////////////////////////////////////////
+// Filter
+///////////////////////////////////////////////////////////////
+
+// ValueOperator is an operator between a field and a value
+type ValueOperator string
+
+// Eq ($eq) checks that field == value
+const eq ValueOperator = "$eq"
+
+// Gt ($gt) checks that field > value
+const gt ValueOperator = "$gt"
+
+// Gte ($gte) checks that field >= value
+const gte ValueOperator = "$gte"
+
+// Lt ($lt) checks that field < value
+const lt ValueOperator = "$lt"
+
+// Lte ($lte) checks that field <= value
+const lte ValueOperator = "$lte"
+
+// LogicOperator is an operator between two filters
+type LogicOperator string
+
+// And ($and) checks that filter && filter2
+const and LogicOperator = "$and"
+
+// Not ($not) checks that !filter
+const not LogicOperator = "$not"
+
+// Or ($or) checks that filter1 || filter2 || ...
+const or LogicOperator = "$or"
+
+// Nor ($nor) checks that !(filter1 || filter2 || ...)
+const nor LogicOperator = "$nor"
+
+// A Filter is a filter on documents, to be passed
+// as the selector of a couchdb.FindRequest
+// In the future, we might add go-side validation
+// but we will need to duplicate the couchdb UCA algorithm
+type Filter interface {
+	ToMango() map[string]interface{}
+	MarshalJSON() ([]byte, error)
```

Comment:
You could inherit the [`json.Marshaler` interface](https://golang.org/pkg/encoding/json/#Marshaler)